File: 0200/0200.go
func numIslands(grid [][]byte) int {
    ans := 0

    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[0]); j++ {
            if grid[i][j] == '1' {
                dfs(grid, i, j)
                ans++
            }
        }
    }

    return ans
}

func dfs(grid [][]byte, i int, j int) {
    if i < 0 || i == len(grid) || j < 0 || j == len(grid[0]) || grid[i][j] != '1' {
        return
    }

    grid[i][j] = '2'
    dfs(grid, i+1, j)
    dfs(grid, i-1, j)
    dfs(grid, i, j+1)
    dfs(grid, i, j-1)
}

File: 0200/0200.js
function numIslands(grid) {
  const ans = 0;

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++) {
      if (grid[i][j] === '1') {
        dfs(grid, i, j);
        ans++;
      }
    }
  }

  return ans;

  function dfs(grid, i, j) {
    if (i < 0 || i === grid.length || j < 0 || j === grid[0].length || grid[i][j] !== '1') {
      return;
    }

    grid[i][j] = '2';
    dfs(grid, i + 1, j);
    dfs(grid, i - 1, j);
    dfs(grid, i, j + 1);
    dfs(grid, i, j - 1);
  }
}

